spinalcordtoolbox/image.py

```diff
@@ -1704,3 +1707,66 @@ def compute_cross_corr_3d(image: Image, coord, xrange=list(range(-10, 10)), xshi
     # Change adjust rl_coord
     logger.info('R-L coordinate adjusted from %s to  %s)', x, x + xrange[ind_peak])
     return x + xrange[ind_peak]
+
+
+def stitch_images(fnames_in: list, fname_out='stitched.nii.gz'):
+    """
+    Stitch two (or more) images utilizing the C++-precompiled binaries of Biomedia-MIRA's stitching toolkit
+    (https://github.com/biomedia-mira/stitching) by placing a system call.
+
+    :param fnames_in: list of filenames of scans in RPI orientation. e.g. [stack1.nii.gz, stack2.nii.gz]
+    :param fname_out: filename for stitched output scan.
+    :param directory: path to temp directory where we store the result of stitching.nii.gz
+    :return: none
+    """
+    # stringify the fname list to parsable cmd parameter
+    fnames_cmd = " ".join(fnames_in)
+
+    cmd = ['isct_stitching', '-i', fnames_cmd, '-o', fname_out, '-a']
+    status, output = run_proc(cmd, verbose='verbose', is_sct_binary=True)
+    print(status, output)
+    if status != 0:
+        raise RuntimeError(f"Subprocess call {cmd} returned non-zero: {output}")
+
+    return Image(fname_out)
+
+
+def generate_stitched_qc_images(ims_in: Sequence[Image], im_out: Image):
+    """
+    Pad input and output images, so that a QC report can compare between the output of
+    'stitching', vs. the naive output of concatenating the input images together.
+    """
+    # find the max shape of all input images
+    shape_max = [max(im.data.shape[0] for im in ims_in),
+                 max(im.data.shape[1] for im in ims_in),
+                 max(im.data.shape[2] for im in ims_in)]
+
+    # pad any input images that are smaller than the max shape
+    # (the stitching tool can handle mismatched image shapes natively, but we have to manage it ourselves)
+    for im in ims_in:
+        for i in [0, 1]:
+            diff = shape_max[i] - im.data.shape[i]
+            if diff:
+                padding = [[0, 0], [0, 0], [0, 0]]
+                padding[i][0] = round(diff / 2)
+                padding[i][1] = diff // 2
+                im.data = np.pad(im.data, padding)
```

Comment:
Two slight problems here:
* This calls `np.pad` twice, once for the x direction and once for the y direction, but a single call could do both at once. Judging by the `if diff:` test, I assume that we care about not doing this (slightly expensive) operation more than necessary.
* The padding computation will be off-by-one in some cases, because `round(diff/2) + diff//2` is not always equal to `diff`:

  ```py3
  >>> [diff for diff in range(4)]
  [0, 1, 2, 3]
  >>> [round(diff/2) + diff//2 for diff in range(4)]
  [0, 0, 2, 3]
  ```
  (The pattern repeats mod 4.)

The following change would fix both issues:
```suggestion
        # the images get concatenated in the z direction,
        # so we only need to pad in the x, y directions
        x_diff = shape_max[0] - im.data.shape[0]
        y_diff = shape_max[1] - im.data.shape[1]
        if (x_diff, y_diff) != (0, 0):
            # note that (diff // 2) + ((diff + 1) // 2) == diff
            im.data = np.pad(im.data, [
                [x_diff // 2, (x_diff + 1) // 2],
                [y_diff // 2, (y_diff + 1) // 2],
                [0, 0],
            ])
```

Reply:
Thank you for being so thorough about this!

> * The padding computation will be off-by-one in some cases, because `round(diff/2) + diff//2` is not always equal to `diff`:

Huh! TIL that [`round(0.5) != 1`](https://stackoverflow.com/a/10825998) in Python3. I'll try to be a bit more careful about rounding assumptions in the future. :) 

Your solution works like a charm, though:

  ```py3
  >>> [(diff // 2) + ((diff + 1) // 2) for diff in range (4)]
  [0, 1, 2, 3]
  ```

I also love the comments you added to explain why we're only padding in the x/y direction.